src/execution/__tests__/executor-test.ts

```diff
@@ -1321,3 +1327,199 @@ describe('Execute: Handles basic execution tasks', () => {
     expect(possibleTypes).to.deep.equal([fooObject]);
   });
 });
+
+describe('Execute: Handles Semantic Nullability', () => {
+  const DeepDataType = new GraphQLObjectType({
+    name: 'DeepDataType',
+    fields: {
+      f: { type: new GraphQLNonNull(GraphQLString) }
+    },
+  });
+
+  const DataType: GraphQLObjectType = new GraphQLObjectType({
+    name: 'DataType',
+    fields: () => ({
+      a: { type: new GraphQLSemanticNullable(GraphQLString) },
+      b: { type: new GraphQLSemanticNonNull(GraphQLString) },
+      c: { type: new GraphQLNonNull(GraphQLString) },
+      d: { type: new GraphQLSemanticNonNull(DeepDataType) }
+    }),
+  });
+
+  it('SemanticNonNull throws error on null without error', async () => {
+    const data = {
+      a: () => 'Apple',
+      b: () => null,
+      c: () => 'Cookie'
+    };
+  
+    const document = parse(`
+      query {
+        b
+      }
+    `);
+
+    const result = await execute({
+      schema: new GraphQLSchema({ query: DataType }),
+      document,
+      rootValue: data,
+    });
+
+    let executable = document.definitions?.values().next().value as ExecutableDefinitionNode;
+    let selectionSet = executable.selectionSet.selections.values().next().value;
```

Comment:
These are just tests so it's not the biggest deal, but is there a better way to retrieve the node for use in error construction? This is workable, but unwieldy.

Reply:
Because we do not yet enable noUncheckedIndexAccess (see https://github.com/graphql/graphql-js/pull/3867) I think you could just index it via 0 at this point although I think what you are doing is preferred, for example

https://github.com/graphql/graphql-js/blob/423e72dec44feb5656a0ea2a384bac68a462732c/src/execution/subscribe.ts#L217

Has been changed in main to your method.

You could assert that you have an executable definition or selection node instead of using typescript casting….

In fact, I think that needs to be improved in main for the example I gave!

https://github.com/graphql/graphql-js/blob/831c1211d7a88bed1758d5c33416514947247241/src/execution/execute.ts#L2241

Reply:
https://github.com/graphql/graphql-js/blob/423e72dec44feb5656a0ea2a384bac68a462732c/src/language/predicates.ts#L24

Reply:
Paid greater attention when working on https://github.com/twof/graphql-js/pull/5 and I think the answer is that generally the tests use `expectJSON` and some other testing deduplication, added second commit there demonstrating use.